src/ModelessDialogHack.mm

```diff
@@ -0,0 +1,69 @@
+#import <objc/runtime.h>
+#import "FinaleHackManager.h"
+#import "FinaleHackList.h"
+#import "FinaleHacksUtils.h"
+
+// Static key for detecting plugin windows.
+static const void *kIsPluginWindowKey = &kIsPluginWindowKey;
+
+void registerModelessDialogHack() {
+  [FinaleHackManager registerHackWithName:@"ModelessDialogHack" block:^{
+    Method canBecomeMainWindowMethod = class_getInstanceMethod([NSWindow class], @selector(canBecomeMainWindow));
+    if (!canBecomeMainWindowMethod) {
+      NSLog(@"ModelessDialogHack Error: Failed to retrieve canBecomeMainWindow selector for NSWindow class. Skipping hack installation.");
+      return;
+    }
+
+    IMP originalImplementation = method_getImplementation(canBecomeMainWindowMethod);
+    if (!originalImplementation) {
+      NSLog(@"ModelessDialogHack Error: Failed to get the implementation of canBecomeMainWindow method. Skipping hack installation.");
+      return;
+    }
+
+    IMP newImplementation = imp_implementationWithBlock(^(id self, SEL _cmd) {
+      BOOL currValue = ((BOOL (*)(id, SEL))originalImplementation)(self, _cmd);            
+      NSWindow* win = (NSWindow*)self;
+      if (currValue && objc_getAssociatedObject(win, kIsPluginWindowKey)) {
+        DEBUG_CODE(NSLog(@"Returning canBecomeMainWindowMethod NO for %@", [win title]));
+        return NO;
+      }
+      return currValue;
+    });
+
+    method_setImplementation(canBecomeMainWindowMethod, newImplementation);
+
+    // Event monitor for NSWindowDidBecomeKeyNotification
+    [[NSNotificationCenter defaultCenter] addObserverForName:NSWindowDidBecomeKeyNotification
```

Comment:
Is there a reason that the computation in the body of the notification handler can't done at IMP time, rather than tagging the window at notification time?  I don't think Apple documents the order of those calls, and this may be paranoia, but I worry that Apple might do some optimization in the future that might reverse the order and break this.  :-)

If you're doing it this way for performance reasons (e.g. if that method gets called frequently), maybe retrieve the associated object and, if it is nil, populate it, thus guaranteeing that the value will be set to either YES or NO exactly once before you use the value.

Thoughts?

Reply:
The main reason is that the JW plugins lose their delegate after they first start up, and the delegate name is the only reliable way to identify them. (That is, the delegate becomes nil after the window is running.) This is probably a bug, but bugs are what we have to live with in the Finale world, lol. I actually had all the code in the IMP and then rediscovered the issue.

My plugin code has been doing it this way for a long time. I don't think the order matters because once the event code runs, the window will not return `YES` for `canBecomeMainWindow` again. Given that both are called during window startup, it doesn't matter for the purposes of what we are trying to achieve.

Reply:
It would be possible to do the whole thing in the IMP anyway. But it feels wrong to modify `win.hidesOnDeactivate` inside the IMP. Which would mean duplicating the code to detect which ones we care about.

Reply:
Just for completeness' sake, I will mention that RGP Lua modeless plugin windows are *not* included in this patch. They have a different delegate name and do not use `FX_Dialog`. But because I was aware of the issue I was able to implement the solution "correctly" for them. That is, their window class inherits from NSWindow and overrides `canBecomeMainWindow`. It also has a more complex handling of `hidesOnDeactivate`.

I tried to sell the Finale team at MakeMusic on implementing this solution system-wide, but they would have none of it. I don't think they ever fully realized just how debilitating and awful this problem is. Nor how painless the fix is.

Reply:
I just reviewed my RGP Lua code and realized I could use `windowController` to id the JW plugins instead of `delegate`. The window controller reliably stays put. That leaves the issue of what to do about `win.hidesOnDeactivate`. Let me know your preference.

Reply:
I'm okay with it either way.